### PR TITLE
AI TermOS Messaging Improvements

### LINKED ISSUE #6407: AI replies to terminal messages are capped to 64 characters
### Labels

[BUG] [MINOR]

### Description

When an AI attempts to respond to a message sent via TermOS, their message will be cut off to be, at most, 64 characters long (for reference, this description is only 188 characters long)

### Steps to reproduce

1. Receive a message from a TermOS console as AI
2. Attempt to reply with a message longer than 64 characters
3. You receive no feedback on your sent message, but if you look at the terminal you replied to, your message has been cut down to 64 characters

### Expected Behavior

Unsure as to what the character limit should be, but messages should be able to be at least 300 characters, similar to the character limit on the terminal command line

### Additional Information & Screenshots

Here on a local server I filled the command line character limit with 0s and sent the message, which rendered as normal in the following image.
image

I then replied to this message, directly copying and pasting said message into the prompt that pops up when you click the blue sender address. As you can see in the terminal, the message cuts out in the middle - after pasting into a .txt document, it turns out to be exactly 64 characters long. This is consistent with when I tested it in a live round.
image

As a quick note, the TermOS command line character limit is 300 characters

### PR DESCRIPTION
**Pull request by Nexusuxen**
## About the PR 

**Main changes**

Received TermOS connection messages to your mainframe (the ones that look like `(0202969f): Hello, world!` and come from someone connecting to your core from a computer terminal) have been significantly improved:
- You will now receive messages while in your shell and be able to reply to them from your shell (you previously could not)
- Updates regarding a terminal's connection (mainly when someone connects or disconnects to/from you) are now properly sent to your AI eye and your shells. This also includes feedback for when you successfully reply to a message.
- A sound will play when you receive a message with a new sound, tone_beep.ogg [(Source)](https://freesound.org/people/pan14/sounds/263133). You will also hear /sound/machines/bweep.ogg from when a terminal connection is established, and /sound/machines/phones/remote_hangup.ogg when a terminal disconnects. The sounds will also play regardless of if you're in your shell or mainframe or AI eye.
- You can now view TermOS conversation history and view active connections with the View Message Log verb, and can directly attempt to reply to a terminal from this list. Each log is timestamped relative to shift time.
- You can mute the sounds that play as a result of TermOS messages, connections, and disconnections. The button for this is in View Message Logs, at the top of the window.
- Fixes #6407 letting AIs send replies longer than 64 characters. Replies now capped to 300 characters to be consistent with the character limit in the computer terminal command lines. (The improved reply feedback also now helps to let you know when your reply has been cut off)

**Other changes:**
- ~~Reworks `ai/show_message()` to now include a check for if the AI is in their shell or not~~
- Adds a new proc to helpers.dm, `formattedShiftTime()`. Returns shift time as a string in hh:mm format. Call with arg `TRUE` to receive time in hh:mm:ss format.
- Adds soundToPlayer() and textToPlayer(), which are AI procs for playing sound and text to the player using playsound_local() and show_text(), regardless of if they're in their mainframe, eye, or a shell
  - TermOS messaging has been adjusted to use these instead of boutput(), replacing this PR's initial use of show_message() (the show_message() proc is now identical to the current master version)

**Planned changes:**
- ~~Add the flag to the relevant playsound procs which makes sounds ignore vacuum-dampening effects because I forgot to do that~~

**Possible but not guaranteed future changes**
- ~~Ability to view active terminal connections at the top of the message log window~~
   - ~~Ideally the netIDs would be clickable just like how they are in the chat log and rest of the message log~~
- ~~Include replies by the AI in the message log~~
- ~~(This one is probably too difficult for me - that or tired brain is intimidated by long lists of args) Add `ai/playsound_local()` to call `playsound_local()` for shell, eye, and mainframe in 1 proc~~ Not quite this but `soundToPlayer()` has been implemented in its place.
- ~~Fix the inability to respond to messages while in a shell.~~
- ~~A button to toggle whether or not sounds will play from messages, connections, and disconnections~~
- Command to entirely disable (or reenable) receiving messages and connections. Active connections would be terminated, and people attempting to connect will receive a reply informing them that messaging has been disabled.

*Note:* All of the crossed out changes above have been implemented. I've decided I won't be doing the final one.

# Images
An example of what a typical message log might look like. The window continues to be interactable even if you swap between mainframe and camera view, or enter or exit a shell.
image

What one's chat feed might look like with the changes (Mrs. Muggles is unfortunately not included in your AI's firmware update):
image
_The final message here only sends when you click on a terminal's ID to reply to it, but it disconnects before you hit enter._

Here's a (poorly made) video which highlights how the sound and chat feed messages work. Again, this works identically to if you're in your mainframe, eye, or a shell.
https://streamable.com/6bmcd8

# Why's this needed? 

As far as I'm aware the code for TermOS messaging for the AI is incredibly old and not at all intended to exist alongside shells and eyes, and just feels clunky and difficult to use. Lack of audio cues when you get a message, as well as the fact that you don't get messages while in a shell, and there's no way to view messages you missed, makes this a very unreliable mechanic that generally isn't worth using all too much. The changes described overhaul this to be Not Shit, for lack of a better word.

Also miscellaneous changes that would probably help out in other places, as well as the aforementioned bugfix.

## Changelog

```changelog
(u)Nexusuxen
(*)TermOS messaging with the AI has been significantly improved. See the PR or minor changelog for more details!
(+)TermOS messaging for AIs now include audio feedback, properly route everything to shells and eye, with a new window to log and view your conversations! Accessed with the 'View Message Log' AI command.
```

### CODE REVIEW

**Comment by Nexusuxen** [comment]
After chatting in imcoder I'll most likely be undoing the `show_message()` change since I seem to have misunderstood the intent of the proc, though I'll double check to make sure. I'll be instead making the `Boutput()` proc pipe to shells and eye and replacing my uses of `show_message()` with it.

**Comment by Nexusuxen** [comment]
Alright, final major changes I wanted to make are in! I suppose feel free to review whenever!

File: code/mob/living/silicon/ai.dm [hunk 1]
```diff
@@ -276,7 +278,25 @@ var/list/ai_emotions = list("Happy" = "ai_happy",\
 		src.eyecam.show_text(message, color, 0, sight_check, allow_corruption, group)
 	return
 
+// For use when you want to send text to the AI player regardless of it's in its mainframe, eye, or shell
+// Calls show_text(), use its syntax
```
Comment (ZeWaka) [inline code/mob/living/silicon/ai.dm:282]:
```suggestion
/// For use when you want to send text to the AI player regardless of it's in its mainframe, eye, or shell
/// Calls show_text(), use its syntax
```

File: code/mob/living/silicon/ai.dm [hunk 2]
```diff
@@ -276,7 +278,25 @@ var/list/ai_emotions = list("Happy" = "ai_happy",\
...
+/mob/living/silicon/ai/proc/textToPlayer(var/message, var/color = "#000000", var/hearing_check = 0, var/sight_check = 0, var/allow_corruption = 0, var/group)
+	if (src.deployed_shell)
+		src.deployed_shell.show_text(message, color, hearing_check, sight_check, allow_corruption, group)
+	else
+		src.show_text(message, color, hearing_check, sight_check, allow_corruption, group)
+	return
 
+// For use when you want to play a sound to the AI player regardless of if it's in mainframe, eye, or shell
+// Calls playsound_local(), use its syntax
```
Comment (ZeWaka) [inline code/mob/living/silicon/ai.dm:291]:
```suggestion
/// For use when you want to play a sound to the AI player regardless of if it's in mainframe, eye, or shell
/// Calls playsound_local(), use its syntax
```

File: code/mob/living/silicon/ai.dm [hunk 3]
```diff
@@ -62,6 +62,8 @@ var/list/ai_emotions = list("Happy" = "ai_happy",\
 	var/net_id = null
 	var/obj/machinery/power/data_terminal/link = null
 	var/list/terminals = list() //Stuff connected to us over the powernet
+	var/messageLog = null
+	var/termMute = FALSE // controls whether or not the ai will hear termos message notifications
```
Comment (ZeWaka) [inline code/mob/living/silicon/ai.dm:66]:
```suggestion
	/// controls whether or not the ai will hear termos message notifications
	var/termMute = FALSE 
```

Comment (ZeWaka) [reply code/mob/living/silicon/ai.dm:66]: https://spacestation13.github.io/DMByExample/hello/comments.html

File: code/mob/living/silicon/ai.dm [hunk 4]
```diff
@@ -276,7 +278,25 @@ var/list/ai_emotions = list("Happy" = "ai_happy",\
...
 	return
 
+// For use when you want to send text to the AI player regardless of it's in its mainframe, eye, or shell
+// Calls show_text(), use its syntax
+/mob/living/silicon/ai/proc/textToPlayer(var/message, var/color = "#000000", var/hearing_check = 0, var/sight_check = 0, var/allow_corruption = 0, var/group)
+	if (src.deployed_shell)
+		src.deployed_shell.show_text(message, color, hearing_check, sight_check, allow_corruption, group)
+	else
+		src.show_text(message, color, hearing_check, sight_check, allow_corruption, group)
+	return
```
Comment (ZeWaka) [inline code/mob/living/silicon/ai.dm:288]: uslesss

File: code/mob/living/silicon/ai.dm [hunk 5]
```diff
@@ -276,7 +278,25 @@ var/list/ai_emotions = list("Happy" = "ai_happy",\
...
+// For use when you want to play a sound to the AI player regardless of if it's in mainframe, eye, or shell
+// Calls playsound_local(), use its syntax
+/mob/living/silicon/ai/proc/soundToPlayer(soundin, vol as num, vary, extrarange as num, pitch = 1, ignore_flag = 0, channel = VOLUME_CHANNEL_GAME, flags = 0)
+	if (deployed_to_eyecam && src.eyecam)
+		src.eyecam.playsound_local(src.eyecam, soundin, vol, vary, extrarange, pitch, ignore_flag, channel, flags)
+	else if (src.deployed_shell)
+		src.deployed_shell.playsound_local(src.deployed_shell, soundin, vol, vary, extrarange, pitch, ignore_flag, channel, flags)
+	else
+		src.playsound_local(src, soundin, vol, vary, extrarange, pitch, ignore_flag, channel, flags)
+	return
```
Comment (ZeWaka) [inline code/mob/living/silicon/ai.dm:299]: uslesss

File: code/mob/living/silicon/ai.dm [hunk 6]
```diff
@@ -642,22 +663,35 @@ var/list/ai_emotions = list("Happy" = "ai_happy",\
...
 		logTheThing("diary", src, null, ": [t]", "say")
+		src.messageLog += "\[[formattedShiftTime(TRUE)]\] <i>Replied to </i><b>[termid]</b><i> with:</i><br>[t]<hr>"
+
+	if (href_list["mute"])
+		src.toggleTermMute()
+
+	if (href_list["refresh"])
+		src.view_messageLog()
 
 	return
```
Comment (ZeWaka) [inline code/mob/living/silicon/ai.dm:696]: uslesss

File: code/mob/living/silicon/ai.dm [hunk 7]
```diff
@@ -1393,6 +1427,33 @@ var/list/ai_emotions = list("Happy" = "ai_happy",\
 
 	src.show_laws(0)
 	return
```
Comment (ZeWaka) [inline code/mob/living/silicon/ai.dm:1429]: uslesss

Comment (Nexusuxen) [reply code/mob/living/silicon/ai.dm:1430]: Sorry, having a hard time understanding what you mean by 'uslesss'. Probably 7 am brain but could you explain a little?

Comment (Nexusuxen) [reply code/mob/living/silicon/ai.dm]: okay ASSUMING you were referring to the return statements (zjd pointed that out), I'm gonna resolve these now, but pls yell at me if that wasn't the intent

**Comment by Nexusuxen** [comment]
Merge conflict resolved.

File: code/mob/living/silicon/ai.dm [hunk 8]
```diff
@@ -1886,8 +1949,12 @@ var/list/ai_emotions = list("Happy" = "ai_happy",\
 			var/message = signal.data["data"]
 			var/rendered = "<span class='game say'><span class='name'><a href='byond://?src=\ref[src];termmsg=[target]'><b>([target]):</b></a></span>"
 			rendered += "<span class='message'> [message]</span></span>"
-
-			src.show_message(rendered, 2)
+			// we need to let the game know that when a log href is clicked, we need to refresh the window
+			var/logAddress = "<span class='game say'><span class='name'><a href='byond://?src=\ref[src];termmsg=[target];refresh=[TRUE]'><b>([target])</b></a></span>"
+			src.messageLog += "\[[formattedShiftTime(TRUE)]\] Sent by: [logAddress]<br><span class='message'> [message]</span></span><hr>"
```
Comment (ZeWaka) [inline code/mob/living/silicon/ai.dm:1954]: this will fail since `messageLog` is null. set in New.

**Review by ZeWaka — CHANGES_REQUESTED** [review]

**Review by ZeWaka — APPROVED** [review]

**Review by virvatuli-gh — CHANGES_REQUESTED** [review]
I've wanted TermOS messaging sounds for a long time, great choices. Remember if you're adding a new sound (in this case `tone_beep.ogg`) it needs to be added to the soundcache as per the audio guidelines. Some nice AI QoL, I like it.

**Review by virvatuli-gh — APPROVED** [review]

**Pull request merged**